onnxruntime/core/session/custom_ops.cc

```diff
@@ -125,18 +125,37 @@ common::Status CreateCustomRegistry(const std::vector<OrtCustomOpDomain*>& op_do
       size_t type_id_counter = 0;
       auto input_count = op->GetInputTypeCount(op);
       for (size_t i = 0; i < input_count; i++) {
+        onnx::OpSchema::FormalParameterOption option = onnx::OpSchema::FormalParameterOption::Single;
+
+        // only since the ORT API version 7 and onwards will users have defined GetInputCharacteristic() and GetOutputCharacteristic()
+        // in their custom op structs.
+        // So, query it ONLY from API version 7 onwards
+        if (op->version >= 7 && op->GetInputCharacteristic(op, i) == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
```

Comment:
Preserve ABI. For older apps trying to use the newer runtime, the older custom op structs compiled into the app will not have the new method defined and so prevent calling the method if the custom op version (which is the same as the API version) is an earlier version.

If someone is trying to re-compile their old app code with the latest header/lib, it won't compile until they define the new required methods. So this only introduces source incompatibility (we don't guarantee source compatibility.)